File: week 03/week 03.R
library(geosphere)  # For distance calculations
library(dplyr)
library(readxl)

# Read the data
df <- read_excel("clinics.xls")

# Convert latitude and longitude to numeric
df$locLat <- as.numeric(df$locLat)
df$locLong <- as.numeric(df$locLong)

# Filter valid latitude and longitude values
df <- df %>%
  filter(locLong >= -180 & locLong <= 180, locLat >= -90 & locLat <= 90) %>%
  filter(!is.na(locLong) & !is.na(locLat))

# Haversine function
haversine <- function(lat1, lon1, lat2, lon2) {
  distHaversine(matrix(c(lon1, lat1), ncol = 2), matrix(c(lon2, lat2), ncol = 2)) * 0.000621371
}

# Measure time for For-loop implementation
start_time <- Sys.time()
haversine_looping <- function(df) {
  distances <- numeric(nrow(df))
  for (i in 1:nrow(df)) {
    distances[i] <- haversine(40.671, -73.985, df$locLat[i], df$locLong[i])
  }
  return(distances)
}
df$distance_loop <- haversine_looping(df)
loop_time <- Sys.time() - start_time

# Measure time for Apply function
start_time <- Sys.time()
df$distance_apply <- apply(df, 1, function(row) {
  haversine(40.671, -73.985, as.numeric(row["locLat"]), as.numeric(row["locLong"]))
})
apply_time <- Sys.time() - start_time

# Measure time for Vectorized implementation
start_time <- Sys.time()
df <- df %>%
  mutate(distance_vectorized = mapply(haversine, 40.671, -73.985, locLat, locLong))
vectorized_time <- Sys.time() - start_time

# Print execution times
execution_times <- data.frame(
  Method = c("For Loop", "Apply", "Vectorized"),
  Time_Seconds = c(loop_time, apply_time, vectorized_time)
)

print(execution_times)

File: week 03/week 03.py
#%%% imports
import numpy as np
import pandas as pd
import cProfile

#%% read in the data
df = pd.read_excel("clinics.xls")
print(df.head())


#%% define the distance computation function
# Define a basic Haversine distance formula
def haversine(lat1, lon1, lat2, lon2):
    MILES = 3959
    lat1, lon1, lat2, lon2 = map(np.deg2rad, [lat1, lon1, lat2, lon2])
    dlat = lat2 - lat1 
    dlon = lon2 - lon1 
    a = np.sin(dlat/2)**2 + np.cos(lat1) * np.cos(lat2) * np.sin(dlon/2)**2
    c = 2 * np.arcsin(np.sqrt(a)) 
    total_miles = MILES * c
    return total_miles

#%% define a function to compute distance, using a for loop
# Define a function to manually loop over all rows and return a series of distances
def haversine_looping(df):
    distance_list = []
    for i in range(0, len(df)):
        d = haversine(40.671, -73.985, df.iloc[i]['locLat'], df.iloc[i]['locLong'])
        distance_list.append(d)
    return distance_list
cProfile.run("df['distance'] = haversine_looping(df)")

#%%% vectorize code by using series and iterrows
# Haversine applied on rows via iteration%%timeit
haversine_series = []
for index, row in df.iterrows():
    haversine_series.append(haversine(40.671, -73.985, row['locLat'], row['locLong']))
#cProfile.run("df['distance'] = haversine_series")
df['distance'] = haversine_series

#%%% Optimize further 

# Timing apply on the Haversine function %%timeit
df['distance'] = df.apply(lambda row: haversine(40.671, -73.985, row['locLat'], row['locLong']), \
                          axis=1)

#%lprun -f haversine df.apply(lambda row: haversine(40.671, -73.985, row['latitude'], row['longitude']), axis=1)
#%%%%timeit
# Vectorized implementation of Haversine applied on Pandas series
df['distance'] = haversine(40.671, -73.985, df['locLat'], df['locLong'])

#%%
cProfile.run("df['distance'] = haversine(40.671, -73.985, df['locLat'], df['locLong'])")

#%%%%timeit
# Vectorized implementation of Haversine applied on NumPy arrays
df['distance'] = haversine(40.671, -73.985, df['locLat'].values, df['locLong'].values)

#%%
cProfile.run("df['distance'] = haversine(40.671, -73.985, df['locLat'].values, df['locLong'].values)")